src/commons/sagas/NativeJsSaga.ts

```diff
@@ -0,0 +1,95 @@
+import { Context } from 'js-slang';
+import { ExceptionError } from 'js-slang/dist/errors/errors';
+import { SagaIterator } from 'redux-saga';
+import { call, put, select, takeEvery } from 'redux-saga/effects';
+import {
+  NATIVE_JS_RUN,
+  NativeJSEvalPayload,
+  NativeJSEvalResult
+} from 'src/features/nativeJS/NativeJSTypes';
+import { store } from 'src/pages/createStore';
+
+import { OverallState } from '../application/ApplicationTypes';
+import { evalNativeJSProgram } from '../nativeJS/NativeJSEval';
+import { actions } from '../utils/ActionsHelper';
+
+const dummyLocation = {
+  start: { line: 0, column: 0 },
+  end: { line: 0, column: 0 }
+};
+
+export function* nativeJsSaga(): SagaIterator {
+  yield takeEvery(NATIVE_JS_RUN, function* (action: ReturnType<typeof actions.nativeJSRun>) {
+    const { workspace, program }: NativeJSEvalPayload = action.payload;
+
+    // Notify workspace & clear REPL
+    yield put(actions.updateWorkspace(workspace, { isRunning: true }));
+    yield put(actions.clearReplOutput(workspace));
+    const context: Context = yield select(
+      (state: OverallState) => state.workspaces[workspace].context
+    );
+    context.errors = [];
+
+    async function evalWrapper() {
```

Comment:
I think it would be good if we can make this a generic wrapper so that we can also wrap `runInContext` as well because currently, long-running source programs would also block UI and not display an indicator (for source programs, I believe it is the "Stop" & "Pause" buttons) that it is running

Perhaps a generic wrapper could be something like this:

```typescript
export async function runWrapper<ReturnType>(
  runFunc: (...args: any[]) => ReturnType,
  ...runFuncArgs: any[]
): Promise<ReturnType> {
  return await new Promise((resolve, _) => setTimeout(() => resolve(runFunc(...runFuncArgs)), 0));
}

// Example usage
const nativeJSResult: NativeJSEvalResult = yield call(runWrapper, evalNativeJSProgram, program);
const sourceResult = call(runWrapper, runInContext, code, context, {
                                            scheduler: 'preemptive',
                                            originalMaxExecTime: execTime,
                                            stepLimit: stepLimit,
                                            throwInfiniteLoops: throwInfiniteLoops,
                                            useSubst: substActiveAndCorrectChapter})
```

What do you think?

Reply:
I agree with this. Making it generic will make more sense.